File: kotlin/src/main/kotlin/io/github/ocirne/cryptopals/set2/Challenge13.kt
package io.github.ocirne.cryptopals.set2

import io.github.ocirne.cryptopals.crypto.AES
import kotlin.random.Random

class Challenge13Profile {

    private val key = Random.nextBytes(AES.BLOCK_SIZE)

    private val aes = AES.ECB(key)

    fun parse(cookie: String): Map<String, String> {
        return cookie.split("&")
            .map { pair -> pair.split("=") }
            .associate { (key, value) -> key to value }
    }

    fun fromEMail(email: String): String {
        val cleanEMail = email.replace("&", "").replace("=", "")
        return "email=${cleanEMail}&uid=10&role=user"
    }

    fun encrypt(emailAddress: String): ByteArray {
        val profile = fromEMail(emailAddress)
        return aes.encrypt(profile.toByteArray())
    }

    fun decrypt(ct: ByteArray): Map<String, String> {
        val plain = String(aes.decrypt(ct))
        return parse(plain)
    }
}

class Challenge13(private val profile: Challenge13Profile) {

    fun makeAdminRole(): Map<String, String> {
        // Block 1: 'email=fofoo@bar.com.....role='
        val roleBlock = profile.encrypt("fofoo@bar.com").sliceArray(0..31)
        // Block 2: 'admin...' with fake padding
        val adminBlock = profile.encrypt("1234567890admin" + "\u000b".repeat(11)).slice(16..31)
        return profile.decrypt(roleBlock + adminBlock)
    }
}

File: python/src/cryptopals/set2/challenge13.py
import secrets

from Crypto.Cipher import AES

from cryptopals.basics import padding, strip_padding


def key_value_parsing(cookie):
    """
    >>> key_value_parsing('foo=bar&baz=qux&zap=zazzle')
    {'foo': 'bar', 'baz': 'qux', 'zap': 'zazzle'}
    """
    return dict(kv.split("=") for kv in cookie.split("&"))


def profile_for(email: str):
    """
    >>> profile_for('foo@bar.com')
    'email=foo@bar.com&uid=10&role=user'
    >>> profile_for('foo@bar.com&role=admin')
    'email=foo@bar.comroleadmin&uid=10&role=user'
    """
    return "email=%s&uid=10&role=user" % email.replace("&", "").replace("=", "")


def encrypt_profile(user_input, key):
    profile = profile_for(user_input).encode()
    aes = AES.new(key, AES.MODE_ECB)  # NOSONAR
    return aes.encrypt(padding(profile))


def decrypt_profile(secret, key):
    aes = AES.new(key, AES.MODE_ECB)  # NOSONAR
    plain = strip_padding(aes.decrypt(secret)).decode()
    return key_value_parsing(plain)


def make_admin_role():
    """
    >>> make_admin_role()
    {'email': 'fofoo@bar.com', 'uid': '10', 'role': 'admin'}
    """
    key = secrets.token_bytes(16)
    # 'email=fofoo@bar.com.....role='
    role_block = encrypt_profile("fofoo@bar.com", key)[:32]
    # 'admin...' with fake padding
    admin_block = encrypt_profile("1234567812admin" + "\x0b" * 11, key)[16:32]
    return decrypt_profile(role_block + admin_block, key)
